design/design-doc-composition-functions-extra-resources.md

```diff
@@ -0,0 +1,1199 @@
+  reading the required information from its `status`.
+- [**ALPHA**] Create an `EnvironmentConfig`, either manually or from another
+  Composition, with the required information and reference it from their
+  Composition via `spec.environment.environmentConfigs`.
+- Use MR-level referencing, if supported for the source and target of the
+  information, being implemented by the provider.
+
+Each of these strategies has its drawbacks:
+- The creation of `ObserveOnly` resources could result in a significant number
+  of resources being created.
+- The use of `EnvironmentConfigs` requires the Composition author to create and
+  maintain a separate resource, which can be cumbersome and error-prone.
+  However, it has to be said that there is value in the clean separation of concerns
+  between the `EnvironmentConfig`'s producer and consumer(s).
+- Using MR-level referencing is not always possible, as it requires explicit
+  provider support, and hardly ever crosses a single provider's (family) boundaries.
+
+As per the Composition Functions' [specification]:
+```text
+A Function MUST NOT assume it is deployed in any particular way, for example
+that it is running as a Kubernetes Pod in the same cluster as Crossplane.
+
+A Function MUST NOT assume it has network access. A Function SHOULD fail
+gracefully if it needs but does not have network access, for example by
+returning a Fatal result.
+```
+
+A clear consequence of the above assumptions is that a Function should not
+assume access to the Kubernetes API. Which means, that letting Functions fetch
+external resources on their own is not an option, Crossplane needs to provide
+them.
+
+With this proposal we aim to provide Functions a way to request additional
+existing Crossplane resources.
+
+## Proposal
+
+We propose to allow Functions to request extra resources through the
+`RunFunctionResponse` and to allow Composition authors to specify what resources
+a Function is authorized to request through the `PipelineStep` struct.
+
+The `PipelineStep` struct would need to be updated as follows:
+```golang
+// A PipelineStep in a Composition Function pipeline.
+type PipelineStep struct {
+	// Step name. Must be unique within its Pipeline.
+	Step string `json:"step"`
+
+	// FunctionRef is a reference to the Composition Function this step should
+	// execute.
+	FunctionRef FunctionReference `json:"functionRef"`
+
+	// Input is an optional, arbitrary Kubernetes resource (i.e. a resource
+	// with an apiVersion and kind) that will be passed to the Composition
+	// Function as the 'input' of its RunFunctionRequest.
+	// +optional
+	// +kubebuilder:pruning:PreserveUnknownFields
+	// +kubebuilder:validation:EmbeddedResource
+	Input *runtime.RawExtension `json:"input,omitempty"`
+
+	// AllowedExtraResources is a list of resources by name or label selector
+	// that this pipeline step is allowed to request.
+	// +optional
+	AllowedExtraResources []ResourceSelector `json:"allowedExtraResources,omitempty"`
```

Comment:
We could postpone implementing this and have a separate discussion about it.

This said, allowing an external component such as Functions, to get arbitrary additional resources through Crossplane, means it's able to escalate it's (usually zero) privileges to those Crossplane has, which feels like a huge increase in scope w.r.t. the current situation where external components can ONLY access resources THEY previously created. 

I see how Observe-Only resources could feel like they already granted the same "escape" capabilities, however I think they are actually not, what external resources you can get are already constrained by the provider's permissions. `provider-kubernetes`'s "ante litteram" Observe-Only resources do allow to do the same IIRC, but they are constrained by the permissions granted to the provider.

I would be fine not implementing this part, but that would mean we should explicitly declare Functions to be inside our trust boundary and therefore users should be even more careful picking trusted Functions only.

Reply:
> We could postpone implementing this and have a separate discussion about it.

I think we should. Even if this does turn out to be something folks need, I suspect the Composition is probably not the right place to configure it.

Consider a case where I write a function, write a Composition that uses that function, then package it as a Configuration and push it to https://marketplace.upbound.io. Then you, a control plane administrator, want to use that Configuration. It's you the control plane administrator, not me the Composition author, who is probably going to have opinions about what resources a function should be allowed to access on your cluster.

If we did want to place constraints around what a function can ask Crossplane to send it, it might make sense to use a control plane scoped allow list [like we do for providers](https://github.com/crossplane/crossplane/blob/0e47c907d53d50a93cd64e6d63523017029e1737/cmd/crossplane/rbac/rbac.go#L75).